src/mmt_models.py

```diff
@@ -0,0 +1,159 @@
+      hidden_dropout_prob=mmt_config.hidden_dropout_prob,
+      attention_probs_dropout_prob=mmt_config.attention_probs_dropout_prob,
+      max_absolute_position_embeddings=mmt_config.max_absolute_position_embeddings,
+      relative_vocab_size=mmt_config.relative_vocab_size,
+      relative_pos_max_distance=mmt_config.relative_pos_max_distance,
+      initializer_range=mmt_config.initializer_range,
+      use_pre_activation_order=mmt_config.use_pre_activation_order,
+      use_one_hot_lookup=mmt_config.use_one_hot_lookup,
+  )
+  return mmt_models.MmtEncoder(**kwargs)
+
+
+def pretrain_model(mmt_config: configs.MmtConfig,
+                   mpp_output_num_classes: int,
+                   patch_embedding_size: int,
+                   return_core_pretrainer_model: bool = False):
+  """Returns model to be used for pre-training.
+  Args:
+    mmt_config: Configuration that defines the core BERT model.
+    mpp_output_num_classes: The number of output classes for patch prediction.
+      it should be (2 ** output_channel_bits ) ** 3.
+    return_core_pretrainer_model: Whether to also return the `MmtPretrainer`
+      object.
+  Returns:
+    A Tuple of
+    (1) MmtPretrainer `with MmtPretrainLossAndMetricLayer`.
+    (2) core Mmt submodel from which to save weights after pretraining.
+    (3) [optional] core `MmtPretrainer` object.
+  """
+  word_ids = tf.keras.layers.Input(
+      shape=(None,),
+      name='word_ids', dtype=tf.int32)
+  segment_ids = tf.keras.layers.Input(
+      shape=(None,),
+      name='segment_ids', dtype=tf.int32)
+  relative_att_ids = tf.keras.layers.Input(
+      shape=(None, None,),
+      name='relative_att_ids', dtype=tf.int32)
+  att_mask = tf.keras.layers.Input(
+      shape=(None, None,),
+      name='att_mask', dtype=tf.int32)
+  patch_embeddings = tf.keras.layers.Input(
+      shape=(None, patch_embedding_size,),
+      name='patch_embeddings',
+      dtype=tf.float32)
+  masked_text_positions = tf.keras.layers.Input(
+      shape=(None,),
+      name='masked_text_positions',
+      dtype=tf.int32)
+  masked_text_label_ids = tf.keras.layers.Input(
+      shape=(None,),
+      name='masked_text_label_ids',
+      dtype=tf.int32)
+  masked_text_label_weights = tf.keras.layers.Input(
+      shape=(None,),
+      name='masked_text_label_weights',
+      dtype=tf.int32)
+  masked_patch_positions = tf.keras.layers.Input(
+      shape=(None,),
+      name='masked_patch_positions',
+      dtype=tf.int32)
+  masked_patch_label_ids = tf.keras.layers.Input(
+      shape=(None,),
+      name='masked_patch_label_ids',
+      dtype=tf.int32)
+  masked_patch_label_weights = tf.keras.layers.Input(
+      shape=(None,),
+      name='masked_patch_label_weights',
+      dtype=tf.int32)
+
+  transformer_encoder = get_transformer_encoder(mmt_config)
+  transformer_encoder._word_embedding_layer.build(word_ids.shape)
+
+  pretrainer_model = mmt_models.MmtPretrainer(
```

Comment:
nit: is "pretrainer" the right name here? I see "pretraining_model" more often. If you prefer "pretrainer", it's fine, but it sounds weird.